merkledag.go

```diff
@@ -7,7 +7,39 @@ import (
 	cid "github.com/ipfs/go-cid"
 )
 
-var ErrNotFound = fmt.Errorf("merkledag: not found")
+// ErrNotFound is used to signal when a Node could not be found. The specific
+// meaning will depend on the DAGService implementation, which may be trying
+// to read nodes locally but also, trying to find them remotely.
+//
+// The Cid field can be filled in to provide additional context.
+type ErrNotFound struct {
+	Cid cid.Cid
+}
+
+// Error implements the error interface and returns a human-readable
+// message for this error.
+func (e ErrNotFound) Error() string {
+	if e.Cid == cid.Undef {
+		return "ipld: node not found"
+	}
+
+	return fmt.Sprintf("ipld: %s not found", e.Cid)
+}
+
+// Is allows to check whether any error is of this ErrNotFound type.
+// Do not use this directly, but rather errors.Is(yourError, ErrNotFound).
+func (e ErrNotFound) Is(err error) bool {
+	switch err.(type) {
+	case ErrNotFound:
+		return true
+	default:
+		return false
+	}
```

Comment:
We could actually do:

```
	var f interface{
		NotFound() bool
	}
	return errors.As(err, &f)
```

This might then catch any target that implements NotFound, regardless of its actual type. So a user might call `errors.Is(ipldErrNotFound, datastoreErrNotFound)` and get  true. I am not sure we want this though, wdty?

Reply:
Thinking about this, we probably don't want to do this. "Random file not found" is not the same as "dag node not found".

Reply:
yeah, it would be mixing the As() an Is() approaches.